blockprod/src/detail/utils.rs

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) 2021-2024 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+
+use chainstate::{
+    chainstate_interface::ChainstateInterface, BlockIndex, GenBlockIndex, NonZeroPoolBalances,
+};
+use chainstate_types::{pos_randomness::PoSRandomness, EpochData};
+use common::{
+    chain::{
+        block::timestamp::{BlockTimestamp, BlockTimestampInternalType},
+        ChainConfig, PoolId,
+    },
+    primitives::{Amount, BlockHeight},
+};
+
+use crate::BlockProductionError;
+
+#[derive(thiserror::Error, Debug, PartialEq, Eq)]
+pub enum PoSAccountingError {
+    #[error("Staker balance retrieval error: {0}")]
+    StakerBalanceRetrievalError(String),
+}
+
+pub fn get_pool_staker_balance(
+    chainstate: &dyn ChainstateInterface,
+    pool_id: &PoolId,
+) -> Result<Amount, BlockProductionError> {
+    let balance = chainstate
+        .get_stake_pool_data(*pool_id)
+        .map_err(|err| {
+            BlockProductionError::ChainstateError(
+                consensus::ChainstateError::StakePoolDataReadError(*pool_id, err.to_string()),
+            )
+        })?
+        .ok_or(BlockProductionError::PoolDataNotFound(*pool_id))?
+        .staker_balance()
+        .map_err(|err| PoSAccountingError::StakerBalanceRetrievalError(err.to_string()))?;
+
+    Ok(balance)
+}
+
+pub fn get_pool_total_balance(
+    chainstate: &dyn ChainstateInterface,
+    pool_id: &PoolId,
+) -> Result<Amount, BlockProductionError> {
+    let pool_balance = chainstate
+        .get_stake_pool_balance(*pool_id)
+        .map_err(|err| {
+            BlockProductionError::ChainstateError(consensus::ChainstateError::PoolBalanceReadError(
+                *pool_id,
+                err.to_string(),
+            ))
+        })?
+        .ok_or(BlockProductionError::PoolBalanceNotFound(*pool_id))?;
+
+    Ok(pool_balance)
+}
+
+pub fn get_pool_balances_at_heights(
+    chainstate: &dyn ChainstateInterface,
+    min_height: BlockHeight,
+    max_height: BlockHeight,
```

Comment:
I vaguely recall something similar was already discussed recently but it seems appropriate to me to use a range here. Conceptually, it would be somewhat similar to [BTreeMap::range](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.range), used to select a range of keys from a map.

Reply:
> I vaguely recall something similar was already discussed recently but it seems appropriate to me to use a range here. Conceptually, it would be somewhat similar to BTreeMap::range, used to select a range of keys from a map.

Well, after I have made this function generic, I can use the `RangeBounds` trait here, just like `BTreeMap::range` does. But the corresponding function on the `ChainstateInterface` level can't be generic, and the "concrete" range types in Rust are not that great IMO, so I'll probably stick to the separate `min_height`/`max_height` there. And using a range just for this function, converting it to `min_height`/`max_height` on the fly, seems redundant to me. So, unless you insist, I'd leave it as is.

P.S. The discussion you are referring to is probably this one - https://github.com/mintlayer/mintlayer-core/pull/1742#discussion_r1589088425 - and I ended up not using a range in that case.